overlays/haskell.nix

```diff
@@ -529,6 +529,13 @@ final: prev: {
                     ++ (args.modules or [])
                     ++ final.lib.optional (args.ghcOverride != null || args.ghc != null)
                         { ghc.package = if args.ghcOverride != null then args.ghcOverride else args.ghc; }
+                    ++ final.lib.optional (args.contentAddressed.enable)
```

Comment:
This is a bit weird? We already pass through `args.modules`, why do this differently?

Reply:
I see, this lets you pass it through directly as an argument instead of putting it in `modules`. I don't think we want that, there's no reason why this should be privileged in that way. Just put it in `modules`.

Reply:
I can't understand how you mean to do that, the `content-addressed` module is imported in `evalProjectModule`, this way the `args` attrset here contains the options definitions (e.g. `contentAddressed.enable = true`). At this point we need to pass again both the module definitions and declarations since this is a different "`evalModules` environment", currently I'm passing the declarations in `package-set.nix`) and here I'm passing the definitions (the only place where I still have access to `args`).

Maybe I'm wrong and I misunderstood something but I tried to follow what I found it was done with `compiler-nix-name`. The only difference I think it's that `compiler-nix-name` is a module option used only for the "validation" `evalModules` in `evalProjectModule` while the `contentAddressed` moduled is used both for validation and the rest.

Reply:
Maybe I should split the module? Keeping two declarations for the "validation" (as for `compiler-nix-name`) in the `cabalProject` and `stackProject` modules and keeping another "real" module separated?

Reply:
Hmm, I may well be confused, I defer to @hamishmack on the finer points of how the module stuff works,

Reply:
I think @michaelpj's suggestion of putting this in `modules` is more natural. I could not push to this branch, so I have made a new one with the change I think he is suggesting #1528.

Reply:
@hamishmack I didn't know this was possible, it looks cleaner and better now.
Are you sure you want to delete the predicate? One can always implement it in its codebase but I thought it could be handy.

However I'm adding your commit to this branch, or do you prefer to move to the new PR?

Reply:
> Are you sure you want to delete the predicate?

The `modules` let you specify the values like three ways.

```
contentAddressed = true;
packages.package-name.contentAddressed = true;
packages.package-name.components.library.contentAddressed = true;
```

I imagine there might be advantages if we could use function to set these.  This is not unique to `contentAddressed` though and I think we can leave it for another PR.  For instance [projectPackagesExes](https://github.com/input-output-hk/cardano-node/blob/43393ac88faa99986ecaa0be409e8b6353f5e9fe/nix/haskell.nix#L163) here is working around the limitations in a slightly clunky way.

I think improving the interface of the `modules` argument is worth considering and perhaps even replacing it if there a are performance issues that could be solved that way see https://github.com/input-output-hk/haskell.nix/issues/1523#issuecomment-1160983831.